File: ruby/view/markedscroller.rb
# Created by Satoshi Nakagawa.
# You can redistribute it and/or modify it under the Ruby's license or the GPL2.

class MarkedScroller < NSScroller
  attr_accessor :dataSource
  
  def drawRect(frame)
    super_drawRect(frame)
    return unless @dataSource
    ary = @dataSource.scroller_markedPosition(self)
    return if ary.empty?
    
    # prepare transform
  	transform = NSAffineTransform.transform
  	width = self.oc_class.scrollerWidthForControlSize(self.controlSize)
  	scale = self.scaleToContentView
  	transform.scaleXBy_yBy(1.0, scale)
  	offset = self.rectForPart(NSScrollerKnobSlot).y
  	transform.translateXBy_yBy(0.0, offset)
  	
  	# make lines
  	knobRect = self.rectForPart(NSScrollerKnob)
  	lines = []
  	prev = nil
  	ary.each do |i|
    	pt = NSPoint.new(0, i)
    	pt = transform.transformPoint(pt)
    	pt.x = pt.x.ceil
    	pt.y = pt.y.ceil + 0.5
    	next if pt == prev
    	prev = pt
    	line = NSBezierPath.bezierPath
    	line.setLineWidth(1.0)
		  line.moveToPoint(pt)
		  line.relativeLineToPoint(NSPoint.new(width, 0))
		  lines << line
		end
    
    # draw lines
  	NSRectClip(self.rectForPart(NSScrollerKnobSlot).inset(3.0, 4.0))
		@dataSource.scroller_markColor(self).set
		lines.each {|i| i.stroke }
		self.drawKnob
  end
  
  def scaleToContentView
  	self.rectForPart(NSScrollerKnobSlot).height / self.superview.contentView.documentRect.height;
  end
end

File: Classes/Views/MarkedScroller.m
// LimeChat is copyrighted free software by Satoshi Nakagawa <psychs AT limechat DOT net>.
// You can redistribute it and/or modify it under the terms of the GPL version 2 (see the file GPL.txt).

#import "MarkedScroller.h"


@implementation MarkedScroller

@synthesize dataSource;
@synthesize markData;

const static int INSET = 3;

+ (BOOL)isCompatibleWithOverlayScrollers
{
	return self == [MarkedScroller class];
}

- (void)updateScroller
{
	self.markData = [dataSource markedScrollerPositions:self];
}

- (void)drawContentInMarkedScroller
{
	if (!dataSource) return;
	if (![dataSource respondsToSelector:@selector(markedScrollerPositions:)]) return;
	if (![dataSource respondsToSelector:@selector(markedScrollerColor:)]) return;
	
	NSScrollView* scrollView = (NSScrollView*)[self superview];
	int contentHeight = [[scrollView contentView] documentRect].size.height;
	if (!markData || !markData.count) return;
	
	//
	// prepare transform
	//
	NSAffineTransform* transform = [NSAffineTransform transform];
	int width = [self rectForPart:NSScrollerKnobSlot].size.width - INSET * 2;
	CGFloat scale = [self rectForPart:NSScrollerKnobSlot].size.height / (CGFloat)contentHeight;
	int offset = [self rectForPart:NSScrollerKnobSlot].origin.y;
	int indent = [self rectForPart:NSScrollerKnobSlot].origin.x + INSET;
	[transform scaleXBy:1 yBy:scale];
	[transform translateXBy:0 yBy:offset];
	
	//
	// make lines
	//
	NSMutableArray* lines = [NSMutableArray array];
	NSPoint prev = NSMakePoint(-1, -1);
	
	for (NSNumber* e in markData) {
		int i = [e intValue];
		NSPoint pt = NSMakePoint(indent, i);
		pt = [transform transformPoint:pt];
		pt.x = ceil(pt.x);
		pt.y = ceil(pt.y) + 0.5;
		if (pt.x == prev.x && pt.y == prev.y) continue;
		prev = pt;
		NSBezierPath* line = [NSBezierPath bezierPath];
		[line setLineWidth:1];
		[line moveToPoint:pt];
		[line relativeLineToPoint:NSMakePoint(width, 0)];
		[lines addObject:line];
	}
	
	//
	// draw lines
	//
	NSColor* color = [dataSource markedScrollerColor:self];
	[color set];
	
	for (NSBezierPath* e in lines) {
		[e stroke];
	}
}

- (void)drawKnob
{
	[self drawContentInMarkedScroller];
	[super drawKnob];
}

@end
